docs/language/informal/enhanced-type-promotion.md

```diff
@@ -0,0 +1,537 @@
+*   A block whose last statement cannot complete normally
+
+Then, given an if statement of the form `if (b) s1 else s2` that is contained in
+some block:
+
+*   If `b` shows that `v` has type T when `b` is true, and `s2` cannot complete
+    normally, then `v` has the type T in the statements after the if statement
+    in the immediately enclosing block. As in:
+
+    ```dart
+    int stringLength(Object o) {
+      if (o is String) /* whatever */; else throw "Not a string.";
+      return o.length; // Promoted here.
+    }
+    ```
+
+*   If `b` shows that `v` has type T when `b` is false, and `v` can be promoted,
+    and `s1` cannot complete normally, then `v` has the type T in the statements
+    after the if statement in the immediately enclosing block.
+
+    This is also allowed if the if statement has no else clause at all.
+
+*Optional: We could expand the definition of complete normally for a block to
+allow for returns or throws before the last statement as long as they are
+unconditionally executed.*
+
+## Promoting More Variables
+
+The specification places several restrictions on the variables whose type can be promoted. The variable `v`:
+
+*   Must be either a local variable or a parameter,
+
+*   cannot be potentially assigned to within the scope S that the promotion
+    would apply to,
+
+*   cannot be potentially assigned to within any closure, and
+
+*   if `v` is captured by a closure in the scope S then `v` cannot be
+    potentially mutated anywhere.
+
+This is a little too restrictive. We relax the restriction on potential
+assignments to allow assignments of the same type, as in:
+
+```dart
+int stringLength(Object o) {
+  if (o is String) {
+    o = o.trim();
+    return o.length;
+  }
+}
+```
+
+Here, since `trim()` returns a String, the assignment of `o` to its result does
+not disable its promotion.
+
+We couldn't relax this restriction before strong mode because the static type of
+expressions could not be trusted. Now that we have a sound type system, it's a
+reasonable change. This does imply, though, that during static analysis of type
+promotion, we need to compute the static type of the right-hand side of
+assignment expressions to be able to determine if promotion applies.
+
+*Optional: The restriction on mutation within closures could be relaxed to only
+apply to closures defined before the boolean expression that allows for the
+promotion or within the scope that the promotion would apply to. A similar
+relaxation could be made for the locations of closures that capture the
+variable. This would require defining precisely what we mean by "before", which
+might be more work than it's worth.*
+
+*Optional: The restriction on mutation within the scope that the promotion would apply to could be relaxed by allowing the promotion to apply to everything up to the point of mutation. For example:*
+
+```dart
+if (v is T) {
+  // v has type T here
+  v = someNonTValue;
+  // v has the non-promoted type here
+}
+```
```

Comment:
I believe the current proposal already covers this case. The wording ("same type") wasn't clear, but the example was analogous to this one.

I updated the wording to be clear that "same type" means the same as the promoted type.

This doesn't, I think, require abstract interpretation or changing the type of a variable after the assignment.